### PR TITLE
Bugfixes for generating dense sketching operators

### PR DESCRIPTION
**Pull request by kaiwenhe7**
This PR resolves the bug from #57 in a way that supports the functionality mentioned in #53.

### CODE REVIEW

**Comment by rileyjmurray** [comment]
Thanks for re-opening the PR! My first comment concerns the tests. Please remove unnecessary print statements and please enable tests that are currently commented-out.

**Comment by rileyjmurray** [comment]
@kaiwenhe7 you recently left a comment on PR #57 mentioning that you've tested matrices whose dimensions are prime numbers. Please push those test changes to this branch so that we can inspect them.

**Comment by rileyjmurray** [comment]
Thanks for that change, @kaiwenhe7. Please also address my comments from earlier:

https://github.com/BallisticLA/RandBLAS/pull/62#issuecomment-1589947904

**Comment by kaiwenhe7** [comment]
@rileyjmurray @TeachRaccooon I am currently working to try and resolve the uncommented tests.

**Comment by TeachRaccooon** [comment]
> @rileyjmurray @TeachRaccooon I am currently working to try and resolve the uncommented tests. Is it okay if I rewrite the concatenation tests just using fill_dense_submat_impl? I don't fully grasp all of the moving parts of fill_dense_submat. I am currently stuck on the debugging, and it may be good for me to simplify it just using fill_dense_submat_impl.

Sure, whatever works!

**Comment by TeachRaccooon** [comment]
@kaiwenhe7 any updates?

**Comment by kaiwenhe7** [comment]
@TeachRaccooon I'm struggling with some things regarding rounding when we're portioning the matrix. I have it working for when the dimensions are even. When we we're splitting the matrix using n_cols / 2, I may be messing up some rounding stuff.

**Comment by kaiwenhe7** [comment]
@TeachRaccooon There is something interesting from doing more testing. I copied the test_identity function to a separate c++ file and when I ran the exact same test, it seems to not have any errors. Here are some screen shots. If any entry of A and B are not equal the values would be printed out, but when running the file, nothing is printed, which means that each entry should be bitwise equal. On the other hand, the exact same function fails the test in the testing suite.

Screen Shot 2023-06-14 at 4 45 05 PM
Screen Shot 2023-06-14 at 4 45 22 PM

**Comment by kaiwenhe7** [comment]
@rileyjmurray I have the truncated version of the submatrix generation implemented along with the exact same tests from the previous implementation. One last choice that needs to be made is the state of the counter than needs to be returned at the end. If we decide to return the state corresponding to the last element of the submatrix, if we want to add rows to the matrix, the user would need to advance to the next counter and then pass it in. Alternatively, we could ourselves by advancing the counter an additional time before returning the state.

**Comment by rileyjmurray** [comment]
Excellent! I'd like for us to advance the counter one additional time automatically, rather than leave that to the user.

**Comment by rileyjmurray** [comment]
(Sorry for accidentally closing the PR for a second there.)

**Comment by kaiwenhe7** [comment]
@rileyjmurray I have the correct state returned and I modified test_identity for the sub matrix concatenation to use the truncated implementation. In the meantime, I will work on the method of returning the correct state that avoids the risk of overflow from the ptr.

**Comment by kaiwenhe7** [comment]
@rileyjmurray Should I delete the old implementations of the submatrix generation?

**Comment by rileyjmurray** [comment]
@kaiwenhe7 if the new implementation is tested at least as thoroughly as the old implementation, then yes.

**Comment by kaiwenhe7** [comment]
@rileyjmurray I made all of the changes you requested in our last meeting. In regards to the counter stuff, maybe that can be done in a separate pull request? Let me know what you think of the current changes.

File: test/test_dense/test_construction.cc [hunk 1]
```diff
@@ -336,3 +337,139 @@ TEST_F(TestFillAxis, long_axis_2x4) {
...
+        RandBLAS_Testing::Util::fill_dense_rmat_trunc<T, r123::Philox4x32,r123ext::boxmul>(B.data(), n_rows, n_cols, state);
+
+        for (int i = 0; i < size; i++) {
+            total += abs(A[i] - B[i]);
+        }
+
+        ASSERT_TRUE(total == 0);
+
+    }*/
+};
```
Comment (rileyjmurray) [inline test/test_dense/test_construction.cc:454]: @kaiwenhe7 is this test commented out deliberately?

**Comment by rileyjmurray** [comment]
@kaiwenhe7 I left a comment about a test that's been commented out. Besides that, I agree that we can leave the delicate counter incrementing to the next PR.

File: test/test_dense/test_construction.cc (hunk 1 above)
Comment (kaiwenhe7) [reply test/test_dense/test_construction.cc:454]: @rileyjmurray Yes. I commented it out because the original test_identity works now using fill_dense.

Comment (rileyjmurray) [reply test/test_dense/test_construction.cc:454]: Deleting it would be the right move here

Comment (kaiwenhe7) [reply test/test_dense/test_construction.cc]: Cool done!

**Pull request merged**